File: src/main/java/io/github/xf8b/xf8bot/commands/EvalCommand.java
/*
 * Copyright (c) 2020 xf8b.
 *
 * This file is part of xf8bot.
 *
 * xf8bot is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * xf8bot is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xf8bot.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.xf8b.xf8bot.commands;

import com.google.common.collect.Range;
import io.github.xf8b.xf8bot.api.commands.AbstractCommand;
import io.github.xf8b.xf8bot.api.commands.CommandFiredEvent;
import io.github.xf8b.xf8bot.api.commands.arguments.StringArgument;
import org.codehaus.groovy.jsr223.GroovyScriptEngineImpl;
import org.jetbrains.annotations.NotNull;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

import javax.script.ScriptEngine;
import javax.script.ScriptException;

public class EvalCommand extends AbstractCommand {
    private static final StringArgument CODE_TO_EVAL = StringArgument.builder()
            .setIndex(Range.atLeast(1))
            .setName("code to eval")
            .build();

    public EvalCommand() {
        super(AbstractCommand.builder()
                .setName("${prefix}eval")
                .setDescription("Evaluates code. Bot administrators only!")
                .setCommandType(CommandType.BOT_ADMINISTRATOR)
                .addAlias("${prefix}evaluate")
                .setMinimumAmountOfArgs(1)
                .addArgument(CODE_TO_EVAL)
                .setBotAdministratorOnly());
    }

    @NotNull
    @Override
    public Mono<Void> onCommandFired(@NotNull CommandFiredEvent event) {
        String thingToEval = event.getValueOfArgument(CODE_TO_EVAL).get();
        ScriptEngine engine = new GroovyScriptEngineImpl();
        try {
            engine.put("event", event);
            engine.put("guild", event.getGuild().block());
            engine.put("channel", event.getChannel().block());
            engine.put("member", event.getMember().get());
            String result = String.valueOf(engine.eval("import discord4j.common.util.Snowflake;" + thingToEval));
            return event.getMessage()
                    .getChannel()
                    .flatMap(messageChannel -> messageChannel.createMessage("Result: " + result))
                    .subscribeOn(Schedulers.boundedElastic())
                    .then();
        } catch (ScriptException exception) {
            return event.getMessage()
                    .getChannel()
                    .flatMap(messageChannel -> messageChannel.createMessage("ScriptException: " + exception))
                    .then();
        }
        //TODO: see if this threading works
    }
}

File: src/main/kotlin/io/github/xf8b/xf8bot/commands/botadministrator/EvalCommand.kt
/*
 * Copyright (c) 2020, 2021 xf8b.
 *
 * This file is part of xf8bot.
 *
 * xf8bot is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * xf8bot is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with xf8bot.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.xf8b.xf8bot.commands.botadministrator

import com.google.common.collect.Range
import io.github.xf8b.xf8bot.api.commands.Command
import io.github.xf8b.xf8bot.api.commands.CommandFiredEvent
import io.github.xf8b.xf8bot.api.commands.arguments.StringArgument
import io.github.xf8b.xf8bot.util.LoggerDelegate
import io.github.xf8b.xf8bot.util.set
import io.github.xf8b.xf8bot.util.toSingletonImmutableList
import kotlinx.coroutines.reactive.awaitSingle
import kotlinx.coroutines.reactor.mono
import org.codehaus.groovy.jsr223.GroovyScriptEngineImpl
import reactor.core.publisher.Mono

class EvalCommand : Command(
    name = "\${prefix}eval",
    description = "Evaluates Groovy code. Bot administrators only!",
    commandType = CommandType.BOT_ADMINISTRATOR,
    aliases = "\${prefix}evaluate".toSingletonImmutableList(),
    arguments = CODE.toSingletonImmutableList(),
    botAdministratorOnly = true
) {
    override fun onCommandFired(event: CommandFiredEvent): Mono<Void> = mono {
        val code = event[CODE]!!
        val engine = GroovyScriptEngineImpl()

        engine["event"] = event
        engine["guild"] = event.guild.awaitSingle()
        engine["channel"] = event.channel.awaitSingle()
        engine["member"] = event.member.get()

        engine.eval(
            """
            import discord4j.common.util.Snowflake;
            $code
            """.trimIndent()
        )?.toString() ?: "no result"
    }
        .flatMap { result -> event.channel.flatMap { channel -> channel.createMessage("Result: $result") } }
        .doOnError { throwable -> LOGGER.error("Error happened during evaluation of code!", throwable) }
        .onErrorResume { throwable -> event.channel.flatMap { channel -> channel.createMessage("$throwable") } }
        .then()

    companion object {
        private val LOGGER by LoggerDelegate()

        private val CODE = StringArgument(
            name = "code",
            index = Range.atLeast(0)
        )
    }
}